Permissions: code cleanup & basic unit tests

ErisDS opened the pull request:
This PR has a LOOOOT of commits but they are all really small.

refs #9043 

These are all non-functional changes. No code or behaviour has changed.

- I have cleanup up our require code, much like I did for the API recently
- I split some of the functions out of the index file, and split then out the tests, which shows they are 100% covered but the core permissions code is not covered
- I added some coverage to the core permissions code

Code coverage currently looks like:
![](https://puu.sh/xFXKp.png)

TODO (in progress - will get coverage to 100%):
- [x] Add tests for "effective.js" 
- [x] Add tests for user/app related code in index.js

Re: #9043, the first section, "High Priority" is covered by this PR, the remainder of the sections will be covered by follow-up PRs or specs.

kirrg001 commented:
Every single change makes sense 👍 
I would merge this PR and you can add your TODO's on top. Maybe update the PR description, that the TODO's will happen in a next PR.

ErisDS commented:
The last 3 commits sort out the remaining test coverage: 
![](https://puu.sh/xGnMa.png)

The first commit fixes an issue I ran into - I wrote the tests for effective.js, they passed in isolation and then didn't pass with `grunt test-unit`, failing on `.should.be.instanceOf(models.Base.Model)`. 

The other 2 add coverage for 2 different areas.

I have found some weird cases. They are tested as-is for now.

When merging, this can be squashed with a short message, E.g:

```
refs #9043

- Split public-related and context code into logical components
- Split tests up to match
- Ensure we have 100% unit test coverage
- General cleanup
```
